#! /bin/sh -e

# DP: Description: Fix segfault when strings contain a mix of forward
#     and backward rules.
# DP: Related bugs: #310635 BZ645
# DP: Dpatch Author: Denis Barbier <barbier@linuxfr.org>
# DP: Patch Author: Denis Barbier
# DP: Upstream status: not yet submitted.
# DP: Test case: the following command segfaults in en_US.UTF-8 locale
# DP:            when BZ645 is fixed:
# DP:                echo 2d d194 0a 2d d194 0a | xxd -r -p | sort
# DP: Date: 2006-03-06

--- libc/string/strcoll_l.c	14 Mar 2004 20:52:47 -0000	1.4
+++ libc/string/strcoll_l.c	23 May 2005 22:35:59 -0000
@@ -370,7 +370,10 @@
 			/* The last pushed character was handled.  Continue
 			   with forward characters.  */
 			if (idx1cnt < idx1max)
-			  idx1now = idx1cnt;
+			  {
+			    idx1now = idx1cnt;
+			    backw1_stop = ~0ul;
+			  }
 			else
 			  {
 			    /* Nothing anymore.  The backward sequence
